Remove our dependency with "make", make it easier to develop on Windows

DanReyLop opened the pull request:
What do you folks think about this? This hasn't taken me a long time, so it's OK if you reject this approach on principle, but I'd want to know before commiting more time to this.

## Rationale
I'm trying to make it easier for Windows contributors (and Automatticians) to get up and running with Calypso. Right now, the options if you are a Windows develooper are:
* Use [Calypso Bootstrap](https://github.com/Automattic/wp-calypso-bootstrap), which is basiclly a Linux VM, and if you want to code on a GUI editor, `rsync` the files between your computer and the VM.
* Go through [these instructions](https://github.com/Automattic/wp-calypso/blob/master/docs/windows.md).

I've tried to follow the instructions a few months ago (or I just tried to install `make` on my own, I don't remember). I couldn't get it to work. Installing `Git` (with a specific configuration), `node` (with a specific configuration), `MSYS2`, play with `pacman` and then finally install `make` seems like a high barrier of entry. Syncing your files with a VM is not ideal either, I remember trying to setup `Unison` and failing too.

Now that we're going to start allowing third-party plugins, we should make the barrier to start developing on Calypso as low as possible.

The only real reason building Calypso isn't fully multi-platform is because of the `Makefile`. Everything else are JS tools, which are perfectly fine in Windows. The thing is, all that `Makefile` does is call JS tools.

## Instructions
I've focused on porting `make run` to NPM scripts. I ignored all the testing, linting, etc rules and focused on getting a single build path running.

Instead of running Calypso using `make run`, use:
```
npm install
npm start
```

The only requirements on Windows are now as easy to get as on UNIX:
1. A working `NodeJS` installation.
2. There's no number 2. Heck, I even run it in `CMD.exe`, which is the worst excuse of a shell.

You should be able to get to a fully working Calypso even in Windows. By "fully working" I mean *fully*, even devdocs, search index, etc.

To be extra-sure that the `Makefile` isn't being used at all (for example by rebuilding the CSS) I've removed it entirely.

## Limitations
What's missing for now:

- [x] `shonkwrap` equivalent, `shonkwrap` is UNIX-only. I generated the `npm-shrinkwrap.json` file using the full `npm shrinkwrap` command.
- [x] Check the node version before building.
- [x] Bundler (for production builds).
- [x] Git Hooks (they are `bash` scripts).
- [x] Utility commands like `translate`, `clean`, `distclean`, `analyze-bundles`, `docker-*`, etc.
- [x] Linting and testing commands.

What's missing and (probably) won't be added:

- The welcome message.
- Checking that the dependencies are installed before building. ~The normal workflow on a node project is to just run `npm install` after checking out a branch.~ I added it to not disrupt people's workflows :)

## Tech used
* [`cross-env`](https://github.com/kentcdodds/cross-env): To make `$ENV_VARIABLES` work on Windows.
* [`globby`](https://github.com/sindresorhus/globby): To parse a glob pattern (`**/test/*.js`) inside a JS script instead of using the UNIX `find` command.
* [`npm-run-all`](https://github.com/mysticatea/npm-run-all): To run several NPM commands in parallel (didn't wanna be slower than `make -j4`).
* [`mkdirp`](https://github.com/substack/node-mkdirp): `mkdir` JS equivalent that works on Windows too.

Also, if you feel like everything looks too cumbersome in the `package.json` file, we could try [`nps`](https://github.com/kentcdodds/nps), which allows us to move the scripts into a separate file which is `JS` instead of `JSON` (and of course, allows comments, descriptions, we could add common env-variable code there, etc).

gziolo commented on package.json:
```diff
@@ -44,6 +44,7 @@
     "cookie-parser": "1.3.2",
     "copy-webpack-plugin": "3.0.1",
     "creditcards": "2.1.2",
+    "cross-env": "4.0.0-beta.0",
```
New dependencies should rather go to `devDependencies`. Unless I miss something.

DanReyLop replied:
If those dependencies are needed to make a `production` bundle, are they `devDependencies`? I always find this confusing on each different project :)

gziolo replied:
@blowery, @gwwar  - I bet it's still dev if it doesn't get into server or client bundles, right?

blowery replied:
if it's required to build the production bundles, it goes in dependencies. we do a `npm install --production` when building our prod bundles. 

usually deps are for things required to run the thing and devDeps are for things required to build the thing. In our case, it's a little muddly.

gziolo commented on package.json:
```diff
@@ -151,7 +155,19 @@
     "test-test:watch": "nodemon -e js,jsx --exec npm run test-test",
     "lint": "bin/run-lint .",
-    "css-lint": "stylelint 'client/**/*.scss' --syntax scss"
+    "css-lint": "stylelint 'client/**/*.scss' --syntax scss",
```
Love it! ❤️

gziolo commented:
Maintaining all those different if statement to make it work with Windows doesn’t scale well. I like this idea a lot.
It might be a good idea to keep existing Make commands aliasing npm scripts to avoid muscle memory errors.
CircleCi fails because of missing `make` command. Example: https://github.com/Automattic/wp-calypso/blob/master/circle.yml#L18.

blowery commented:
One big thing we lose by dropping `make`: automatically re-running `npm install` when `package.json` changes when switching branches. If we go this road, it's much more likely that folks will switch branches and be running against the "wrong" package.json, which may lead to really odd bugs. I don't know that it's a giant deal, but it's a thing.

DanReyLop commented:
> One big thing we lose by dropping make: automatically re-running npm install when package.json changes when switching branches. If we go this road, it's much more likely that folks will switch branches and be running against the "wrong" package.json, which may lead to really odd bugs. I don't know that it's a giant deal, but it's a thing.

One possible way to handle this is to leave a "stripped-down" `Makefile` with the usual commands (`run`, build for Docker, test/lint, etc) that just call the new `package.json` commands. We could add the `npm install` like this:
```
_install: package.json
    npm install
run: _install
    npm start
lint: _install
    npm run lint
# and so on
```
Or something similar like that.

99% of Node developers (super-accurate percentage, but my sources are confidential) are used to looking on `package.json` for commands to execute `npm run whatever`, and they're also used to do `npm install` when they switch branches (or, at least, that's the first thing to try if something fails). _But_, I don't want to make life of current automatticians _worse_ with this PR, so having feature-parity with the current functionality of the `Makefile` is important, and so is keeping it there to not piss off people who have `make run` in their muscle memory :)

gziolo commented:
There are also `pre*` and `post*` commands available. So you can setup `prestart` or `pretest` to run `npm install`. See https://docs.npmjs.com/misc/scripts.

DanReyLop commented:
Running `npm install` on `wp-calypso`, even if the project is up-to-date, takes a good 10 seconds on a fast machine. I would prefer not to do it before every "main" command if it's not needed. The `Makefile` used a cool optimisation, reading the "last modified timestamp" of the `package.json` file, ideally I would like to replicate something like that.

`yarn check --verify-tree` is a better alternative, it takes only 2-3 seconds, but we're not using `yarn` yet, so 🤷‍♂️

DanReyLop commented on package.json:
```diff
@@ -61,11 +63,12 @@
     "he": "0.5.0",
     "html-loader": "0.4.0",
-    "i18n-calypso": "1.7.0",
+    "i18n-calypso": "Automattic/i18n-calypso#add/cli-globs",
```
Waiting for https://github.com/Automattic/i18n-calypso/pull/36 to be merged and released.

DanReyLop commented on package.json:
```diff
@@ -172,9 +222,10 @@
     "lodash-deep": "1.5.3",
     "md5-file": "3.1.0",
-    "mixedindentlint": "1.1.1",
+    "mixedindentlint": "DanReyLop/mixedindentlint#add/cli-globs",
```
Waiting for https://github.com/sirbrillig/mixedindentlint/pull/4 to be merged and released.

DanReyLop commented:
I believe this is ready for review now. All the `make` tasks have a corresponding, Windows-compatible `npm` task.

Things to note:
* I've added a `npm run install-if-needed` task, which contains a similar optimisation as the previous `make install`.
* I'm waiting for 2 PRs to `i18n-calypso` and `mixedindentlint`, but that doesn't change the bulk of the review.
* I've changed all the references to `make` I could find, and updated the installation instructions (no more Windows-specific instructions yay!).
* I've left a "skeleton" `Makefile` with all the commands, just pointing to the corresponding `npm` task.

DanReyLop commented on package.json:
```diff
@@ -143,15 +148,60 @@
+    "bundle": "npm run env -- node server/bundler/bin/bundler.js",
+    "update-deps": "npm run rm -- node_modules && npm run rm -- npm-shrinkwrap.json && npm install --no-optional && npm install --no-optional && npm shrinkwrap --dev",
+    "preshrinkwrap": "npm prune",
+    "postshrinkwrap": "node bin/trim-shrinkwrap.js",
```
This will be automatically executed every time `npm shrinkwrap --dev` is called, so no need to worry about custom commands or `shonkwrap` anymore :)

DanReyLop commented on package.json:
```diff
@@ -143,15 +148,60 @@
+    "postshrinkwrap": "node bin/trim-shrinkwrap.js",
+    "dashboard": "webpack-dashboard -- npm start",
+    "precommit": "npm run install-if-needed && node bin/pre-commit-hook.js",
+    "prepush": "npm run install-if-needed && node bin/pre-push-hook.js",
```
I've added `husky` to our dependencies: https://github.com/typicode/husky

gwwar commented:
I think this is a good step forward, but could we perhaps split this out into smaller PRs ( a few npm script conversions for each)? This would make testing much easier so we can see if there are any regressions/note the differences in script behavior.

DanReyLop commented:
After talking with @gwwar, I'm not going to split this PR, but at least I'm going to re-do the commits so it's easier to review (the changes to the documentation add A LOT of noise, apart from that the PR is not so big).

This PR will still be quite risky, so I'll put up a detailed test plan. Reverting to "In Progress" while I do that.

DanReyLop commented:
I've split the PR into more reasonable commits. The bulk of the PR is here: https://github.com/Automattic/wp-calypso/pull/12583/commits/c805fceee8609ae30b02de2f909fd08e8a994842

The rest of the commits are documentation changes, some cleanup, etc. It's still a big PR, but at least now it's not so noisy to review.

# Test plan

## Windows
In plain Windows (plain `CMD.exe`, no `Cygwin`, no `bash.exe`, nothing), nothing worked before. So, scrictly speaking, there can't be any regressions with this PR. But of course, the goal of this PR is to make it work.  If you have a Windows machine, try running some of the `UNIX` test cases in your machine, and put your name, case run, and notes in the list below:
* ...

## UNIX
For UNIX we need to watch out for regressions. Please check the box on each test case you run on your machine, and sign off with your user name.

Every reviewer should focus on the area they know more deeply. It's not enough to test that everything appears to work, it's important to dive into the code changes because the scope of these changes mean that a lot of things could be broken in subtle ways.

### Running
- [ ] Run `npm start`, check that Calypso loads on `calypso.localhost:3000`.
- [ ] Navigate to a different section that's lazy loaded, like `post-editor`.
- [ ] Navigate to the auto-generated `devdocs`, check that you can navigate through the `.md` files.
- [ ] Check that the `devdocs` search works.
- [ ] Stop the process, run `npm run clean`, check that the `build/`, `public/*.js`, `public/*.js` and generated devdocs artifacts are gone.
- [ ] Run `npm start` again, check that everything was generated again correctly.
- [ ] Run `npm run distclean`, check that all the files that were removed with `npm run clean` are removed again, plud the folder `node_modules`.
- [ ] Run `npm start` again, check that the dependencies are automatically installed before running the task.
- [ ] Disable some Calypso features using `DISABLE_FEATURES=... npm start`, check it works as expected.

### Building
- [ ] Run `npm run build-desktop`, check that the Desktop build is correct.
- [ ] Run `CALYPSO_ENV=production npm build`, check that the JS modules were correctly minified.
- [ ] Try with other combinations of `CALYPSO_ENV` and `NODE_ENV`.
- [ ] Run `npm run translate`, check the `calypso-strings.pot` file is generated correctly.
- [ ] Change a string in the code, run `npm run translate` again, check that string is in the new `calypso-strings.pot`.

### Testing
- [x] Check that CircleCI succeds. @danreylop
- [ ] Run `npm test` locally, see that it passes.
- [ ] Run `npm lint` or any of the `lint:*` commands, check that they work correctly. Some of them produce tons of errors, but that's expected. Try to force some lint errors to see that they are caught.

### Misc
- [ ] Run `npm run build-docker` to setup a Docker container.
- [ ] Run `npm run docker` to spin up the Docker container.
- [ ] Remove the file `npm-shrinkwrap.json`, run `npm shrinkwrap --dev`, check it's regenerated correctly.
- [ ] Install a dummy package using `npm install left-pad --save`, check that the `npm-shrinkwrap.json` file is correctly updated.
- [ ] Run `npm run update-deps`, check that all the dependencies are reinstalled and a new `npm-shrinkwrap.json` file is generated from scratch.
- [ ] Make a dummy change in `client/*.js` that will trigger a `eslint` error, try to commit, see that it's rejected.
- [ ] Run `npm run analyze-bundles`, check that it works as expected.
- [ ] Run `npm run dashboard`, check that it works as expected.
- [ ] Run some `make *` commands, to check that they still work.

gwwar commented:
For folks looking at the docker commands see also PCYsg-5XR-p2 for instructions on how to generate the secrets file and for updating the hosts file for testing.

gwwar commented:
@DanReyLop could we add another checkbox for testing this with Desktop?

gwwar commented:
;) We're missing the Calypso Ascii art on startup.

![startup](https://cloud.githubusercontent.com/assets/1270189/24880827/b50a9cca-1df0-11e7-8ec5-7677399a4028.png)

gwwar commented:
If we have a mismatching Node version, current behavior is to warn you, but not throw an error. This was requested by folks in the past, so I think we'd want to keep that behavior here.

For reference this is what we see in master:
![warning](https://cloud.githubusercontent.com/assets/1270189/24881145/b63fdb86-1df1-11e7-8e89-33391ff982bc.png)

And this is what it looks like on this branch
![error](https://cloud.githubusercontent.com/assets/1270189/24881159/c14b740e-1df1-11e7-8acf-a833ed55a1a8.png)

DanReyLop commented:
> ;) We're missing the Calypso Ascii art on startup.

Sorry about that, but I only tried to implement it for a grand total of about 10 seconds :D

> If we have a mismatching Node version, current behavior is to warn you, but not throw an error. This was requested by folks in the past, so I think we'd want to keep that behavior here.

Done, I didn't know it was like that on purpose. The proble is that the `npm` output is a lot noisier than `make` (at least by default) so a warning may get unnoticed.

I tried adding `loglevel=silent` to `.npmrc`, but that means that if you do for example `npm install asdfghj`, you won't get any feedback that the package couldn't be found. I guess we could add the `-s` flag to all the `npm` usages in `package.json`?

gwwar commented:
> I guess we could add the -s flag to all the npm usages in package.json?

Yeah, I think that would help reduce the noise. We can always add another alias without the silent flag to use for debugging. How about adding a custom error message in prestart with something like:
```bash
    "prestart": "npm run install-if-needed && check-node-version --package || ( echo 'Node version mismatch detected. Calypso requires the above version. Please switch using nvm or n!' )",
```

gwwar commented:
Getting some not found warnings for devdoc search.

![screen shot 2017-04-10 at 2 13 06 pm](https://cloud.githubusercontent.com/assets/1270189/24882963/d9147760-1df7-11e7-8143-0ac9e44428cb.png)

DanReyLop commented:
> Yeah, I think that would help reduce the noise. (Edit: we'd probably only need to do this for the start command). We can always add another alias without the silent flag to use for debugging.

I've gone ahead and added `loglevel = silent` to the `.npmrc`. It wouldn't be enough to add `--silent` to `npm start`, because even though that particular command would be silent, all the `npm run *` that happen next would still be noisy. That's specially ridiculous when some command fails down the chain, then the `ERR` output can be 80+ lines long, with useless tips to "Contact the owner of the repo".

The only downside I see is that issue with built-in commands such as `npm install`. With `loglevel=silent`, if a `npm install blah` fails, it doesn't provide any output, but if it succeds, it displays the newly installed module(s) tree. Do you think that's enough feedback?

> How about adding a custom error message in prestart with something like

The `check-node-version` output already provides enough context IMO:
```
node: 6.10.0
npm: 3.10.10
yarn: 0.21.3
Error: Wanted node version 6.10.2 (6.10.2)
To install node, run `nvm install 6.10.2` or see https://nodejs.org/
```

> Getting some not found warnings for devdoc search.

My bad, that's caused because `sh` does its own (different) pattern expansion before passing the list of files to the JS script. `CMD.exe` doesn't do any of that, so it passes the glob patterns directly to the JS script, that's why I hadn't noticed. I've fixed it so `sh` doesn't pre-parse the file patterns.

For more context: http://stackoverflow.com/questions/32017169/npm-glob-pattern-not-matching-subdirectories

DanReyLop commented:
Thanks matticbot :) I've just rebased.

DanReyLop commented:
Could I please get a pair of 👀 in this? It's a big PR, and "letting it rot" doesn't seem like a good idea (will be hard to pick up even with constant rebases).

Also, these PRs would need to be merged into their respective repos:
https://github.com/Automattic/i18n-calypso/pull/36
https://github.com/sirbrillig/mixedindentlint/pull/4

gziolo commented on package.json:
```diff
@@ -143,19 +148,65 @@
+    "docker": "docker run -it --name wp-calypso --rm -p 80:3000 -e NODE_ENV='wpcalypso' -e CALYPSO_ENV='wpcalypso' wp-calypso",
+    "rm": "node bin/rm.js",
+    "clean": "run-p clean:*",
+    "clean:public": "npm run rm -- public .css .css.map .js",
```
We need to add `.js.map` to the list. Those files are created when executing `make analyze-bundles` atm. I have another patch open where I add this change.

DanReyLop replied:
Done (and rebased).

stephanethomas commented:
@DanReyLop: a heads-up that I set some time aside next week to review your pull request, as well as to test it on Windows.

gziolo commented:
It would be also nice to rebase with master and make sure we didn't introduce any regression. For instance, I added recently this line https://github.com/Automattic/wp-calypso/blob/master/webpack.config.node.js#L126 in both webpack config files. It seems like it needs to be adjusted a bit.

DanReyLop commented:
> It would be also nice to rebase with master and make sure we didn't introduce any regression. For instance, I added recently this line https://github.com/Automattic/wp-calypso/blob/master/webpack.config.node.js#L126 in both webpack config files. It seems like it needs to be adjusted a bit.

Thanks for the heads-up! I've just rebased and changed that line (in both configs). Too bad those kind of aditions don't give merge conflicts.

gziolo commented on package.json:
```diff
@@ -143,19 +148,65 @@
     "npm": "3.10.10"
   },
   "scripts": {
```
Can you order all options by name alphabetically?

DanReyLop replied:
I've just done it, but I'm not sure what's better: alphabetically, or "grouped" by workflow, similarly how the `Makefile` was structured.

stephanethomas commented:
I tested this pull request on Ubuntu (via Calypso Bootstrap) but was presented with strange `Not found:` labels on the devdocs `Search` page once logged in:

<img width="625" alt="screenshot" src="https://cloud.githubusercontent.com/assets/594356/25584648/276e3d92-2e97-11e7-96e1-0246db13a17a.png">

It took me while but I figured out the culprit, which operates when generating `server/devdocs/search-index.js`. Globbing is indeed a [shell built-in](http://manpages.ubuntu.com/manpages/zesty/man7/glob.7.html) and so any path argument of the following command will be expanded *before* being passed to the `generate-devdocs-index` script:

```
node server/devdocs/bin/generate-devdocs-index \"**/*.md\" \".github/**.md\" \"!build/**\" \"!node_modules/**\" \"!public/**\"
```

I've added a commit to fix that by simply [escaping the double asteriks](https://askubuntu.com/questions/828126/is-globbing-a-feature-of-the-shell).

DanReyLop commented:
> It took me while but I figured out the culprit, which operates when generating server/devdocs/search-index.js. Globbing is indeed a shell built-in and so any path argument of the following command will be expanded before being passed to the generate-devdocs-index script:

I had that exact same problem and I "solved it" quoting the arguments. It worked in Windows' `Git Bash`...

> I've added a commit to fix that by simply escaping the double asteriks.

That breaks it in Windows CMD :(

I'm trying to find a solution that works for all shells.

stephanethomas commented:
I tested this pull request on Windows and it works great!

I'm still concerned by the size of this pull request, and think it would be better to split it in smaller ones. We know it works and we know that's a direction we want to follow. I would rather migrate our code gradually to make sure that we don't break Calypso in the process. I'm happy to help with the review of those pull requests though!

stephanethomas commented:
> That breaks it in Windows CMD :(

Arf, I didn't tested my commit on Windows :(.

gziolo commented:
@danreylop can we add all package.json changes and dependant scripts in one PR and do cleanup in another PR a few days later? I think that would make sure that we have enough people testing it and backup plan when something goes wrong :)

DanReyLop commented:
> can we add all package.json changes and dependant scripts in one PR and do cleanup in another PR a few days later? I think that would make sure that we have enough people testing it and backup plan when something goes wrong :)

Done: #13536 

I have the "cleanup" branch ready too, so as long as the first one is merged, I can put the cleanup PR for review.

gziolo commented:
Many thanks for opening #13536. This should make the whole process much easier 😄

gziolo commented:
Do you plan to rebase it with master and reopen to land rest of the changes in the near future?

DanReyLop commented:
@gziolo I have another branch ready to go with those changes (documentation changes and make the `Makefile` be just a bunch of alias), I'm waiting for a while (a few days at least) to see if there's any problem with the `npm` changes that are already merged. I'll open another PR sometime next week.

gziolo commented:
Cool, thanks 👍
